Using streaming conformer as transducer encoder

pkufool opened the pull request:
I slightly refactored https://github.com/k2-fsa/icefall/pull/242, and create this new pull request.  I think it is no need to duplicate a new folder, as making conformer to streaming just needs very minor fixes.

~~I also add the symbol delay penalty to this PR, see https://github.com/k2-fsa/k2/issues/955 for how we penalize and measure symbol delay.~~

Will post the experiment results here soon.

pkufool commented on egs/librispeech/ASR/pruned_transducer_stateless2/conformer.py:
```diff
@@ -117,17 +151,148 @@ def forward(
+                self.causal
+            ), "Causal convolution is required for streaming conformer."
+            max_len = x.size(0)
+            chunk_size = torch.randint(1, max_len, (1,)).item()
```
The method to choose `chunk_size` here is from wenet's paper, we may tune it to find a better one in the future.

pkufool commented on egs/librispeech/ASR/pruned_transducer_stateless2/conformer.py:
```diff
@@ -117,17 +151,148 @@ def forward(
+        chunk_size: int = 16,
+        left_context: int = 64,
+        simulate_streaming: bool = False,
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
```
The streaming forword interface is emformer-like, it handles decoding states (i.e. caches) outside, which will make it easer to implement async decoding.

danpovey commented:
Cool!
One thing we should perhaps discuss is whether this belongs as a change to pruned_transducer_stateless2, vs.  pruned_transducer_stateless4 which I believe is the latest recipe?  @csukuangfj what are the changes from 2 to 4?

yaozengwei commented:
> Cool! One thing we should perhaps discuss is whether this belongs as a change to pruned_transducer_stateless2, vs. pruned_transducer_stateless4 which I believe is the latest recipe? @csukuangfj what are the changes from 2 to 4?

`pruned_transducer_stateless4` (see https://github.com/k2-fsa/icefall/pull/344) supports saving averaged model and the epoch number counts from 1.

csukuangfj commented:
>  @csukuangfj what are the changes from 2 to 4?

"4" has the changes from zengwei that supports model averaging that you proposed in https://github.com/k2-fsa/icefall/issues/337

Otherwise, 4 is identical to 2. I would suggest to always use 4 as zengwei has showed it is helpful.

From my results of deeper and narrower conformer, it is also helpful.

danpovey commented:
@csukuangfj BTW, are we recommending the deeper and narrower conformer right now?  How much better was it, again?

danpovey commented:
@pkufool it may be fairly easy to use `git diff` to move your changes from 2 to 4.  E.g.
cd pruned_transducer_stateless2 
git diff master..HEAD . > ~/diff
cd ../pruned_transducer_stateless4
patch -p4 < ~/diff
(the 4 may need to be changed, e.g. to 3 or 5, if this fails).

csukuangfj commented:
> @csukuangfj BTW, are we recommending the deeper and narrower conformer right now? How much better was it, again?

I am updating the results and uploading the pre-trained models to huggingface. Please wait for a moment.

pkufool commented:
> Cool! One thing we should perhaps discuss is whether this belongs as a change to pruned_transducer_stateless2, vs. pruned_transducer_stateless4 which I believe is the latest recipe? @csukuangfj what are the changes from 2 to 4?

I think they share the same conformer.py, only a little changes on train.py & decode.py. 

> @pkufool it may be fairly easy to use git diff to move your changes from 2 to 4. E.g.
cd pruned_transducer_stateless2
git diff master..HEAD . > ~/diff
cd ../pruned_transducer_stateless4
patch -p4 < ~/diff
(the 4 may need to be changed, e.g. to 3 or 5, if this fails).

Thanks, I will try it.

pkufool commented:
## Summarize my experiment results for streaming conformer below:
## About the dynamic chunk training
I tried several settings(Note: The chunk sizes and contexts are all in frames after subsampling):
* The first row is the results reproduced with https://github.com/k2-fsa/icefall/pull/242 .
* The second row  was trained with limited left context, because we can only see limited left context for parallel streaming decoding (Note: the left context is in chunks, and the chunk size is dynamicly sampled from [1, 25]). 
* The third and forth rows were trained with fixed chunk size, I tried 16 and 8.
* The fifth row was trained with fixed left context, I only tried left context = 32.

Comparing with the first and second row,  we can see if we need to decode with limited left context, we have to trained with limited left context. Comparing with the second row and last three rows, we can see the models trained with fixed chunk size and fixed left context do not outperform the model trained with dynamic chunk size.
<img width="786" alt="image" src="https://user-images.githubusercontent.com/11765074/174634761-f185c7ab-18cf-4bf8-88ff-8a60413276c4.png">

### About using different left-context for different layers
I also tried train the conformer encoder using different left context for different layers. (The models are all 12 layers conformer).
* The baseline system was trained with num-left-chunks=4 for all layers and decoded with chunk size =16, left-context=64.
* For the second row, the lower 8 layers used num-left-chunks=2, the upper 4 layers used num-left-chunks=4. In decoding time, they also used different left context(32 for lower 8 layers, 64 for upper 4 layers).
* The third row is the same as second row, the only differnece is I splited the layers into 6 + 6.

We can see that the WERs are close to each other for all three settings, smaller left context can save more memories when decoding, though. I did not do much experiments on this idea(i.e. how to split layers, how many left context needed by different layers), so the conclusion might not soild. 
<img width="748" alt="image" src="https://user-images.githubusercontent.com/11765074/174418599-678e61a4-06e8-4bed-b839-1c683be39744.png">

### About streaming decoding and right-context
I also implemented the real paralell streaming decoding for the conformer model, you can see the code in `streaming_decode.py` in each directory. The table below shows the decoding results of several models using different chunk sizes and contexts.  You can see in the table, adding right context usually helps to decrease the WERs, but for some settings, like decoding with left-context=32 and chunk-size=8 in pruned_transducer_stateless2, pruned_transducer_stateless3, adding right context makes it worse. I still have not figured out the reasons.

<img width="782" alt="image" src="https://user-images.githubusercontent.com/11765074/174637078-8983c351-55c5-4c0b-b585-79ad8d9fac38.png">

**Note:** You can also try these streaming conformer models in our server framework Sherpa (https://github.com/k2-fsa/sherpa/tree/master/sherpa/bin/streaming_conformer_rnnt).

csukuangfj commented on egs/librispeech/ASR/pruned_transducer_stateless/decode.py:
```diff
@@ -250,6 +262,69 @@ def get_parser():
         Used only when --decoding_method is greedy_search""",
     )
 
+    parser.add_argument(
+        "--dynamic-chunk-training",
+        type=str2bool,
+        default=False,
+        help="""Whether to use dynamic_chunk_training, if you want a streaming
+        model, this requires to be True.
+        Note: not needed for decoding, adding it here to construct transducer model,
+              as we reuse the code in train.py.
+        """,
+    )
+
+    parser.add_argument(
+        "--short-chunk-size",
+        type=int,
+        default=25,
+        help="""Chunk length of dynamic training, the chunk size would be either
+        max sequence length of current batch or uniformly sampled from (1, short_chunk_size).
+        Note: not needed for decoding, adding it here to construct transducer model,
+              as we reuse the code in train.py.
+        """,
+    )
+
```
You can move these options to train.py and import them into here.
See
https://github.com/k2-fsa/icefall/blob/dc89b61b808a22b64e7e38cb09771ea7a6bb64d2/egs/librispeech/ASR/pruned_transducer_stateless5/train.py#L331

csukuangfj commented on egs/librispeech/ASR/pruned_transducer_stateless/decode.py:
```diff
@@ -301,9 +376,19 @@ def decode_one_batch(
+        encoder_out, encoder_out_lens, _ = model.encoder.streaming_forward(
+            x=feature,
+            x_lens=feature_lens,
+            states=[],
```
I would suggest supporting `None` and not using an empty list `[]` here.

csukuangfj commented on egs/librispeech/ASR/pruned_transducer_stateless/decode.py:
```diff
@@ -250,6 +262,69 @@ def get_parser():
+        """,
+    )
+
+    parser.add_argument(
```
I recommend sharing the parameters from train.py whenever possible.

csukuangfj commented on egs/librispeech/ASR/pruned_transducer_stateless/decode.py:
```diff
@@ -301,9 +376,19 @@ def decode_one_batch(
+            states=[],
+            chunk_size=params.decode_chunk_size,
+            left_context=params.left_context,
+            simulate_streaming=True,
```
When using `simulate_streaming=False,` does it have any difference with the `else` part?

luomingshuang commented on egs/librispeech/ASR/pruned_transducer_stateless2/train.py:
```diff
@@ -40,6 +40,18 @@
+  --exp-dir pruned_transducer_stateless/exp \
+  --full-libri 1 \
+  --dynamic-chunk-training 1 \
+  --causal_convolution 1 \
```
causal_convolution-->causal-convolution

pkufool replied:
Fixed, Thanks!

luomingshuang commented on egs/librispeech/ASR/pruned_transducer_stateless/streaming_decode.py:
```diff
@@ -0,0 +1,679 @@
+    cuts: CutSet,
+    params: AttributeDict,
+    model: nn.Module,
+    sp: spm.SentencePieceProcessor,
```
It seems that this parameter `sp` is not used in this function.

luomingshuang commented on egs/librispeech/ASR/pruned_transducer_stateless/streaming_decode.py:
```diff
@@ -0,0 +1,679 @@
+def decode_one_chunk(
+    params: AttributeDict,
+    model: nn.Module,
+    sp: spm.SentencePieceProcessor,
```
It seems that the parameter `sp` is not used in this function.

pkufool commented:
Merging now. Most of the code added is to support streaming decoding, it won't break any previous recipes. Will fix the issues later in other PRs if there is any.
